hack/generator/pkg/astmodel/storage_conversion_factories.go

```diff
@@ -0,0 +1,360 @@
+		return nil
+	}
+
+	srcPrim := AsPrimitiveType(sourceEndpoint.Type())
+	dstPrim := AsPrimitiveType(dstOpt)
+
+	if srcPrim == nil || dstPrim == nil || !srcPrim.Equals(dstPrim) {
+		// Either or both sides are not primitive types, or not the same primitive type
+		return nil
+	}
+
+	return func(reader dst.Expr, writer dst.Expr, ctx *CodeGenerationContext) []dst.Stmt {
+		return []dst.Stmt{
+			astbuilder.SimpleAssignment(writer, token.ASSIGN, astbuilder.AddrOf(reader)),
+		}
+	}
+}
+
+// assignPrimitiveTypeFromOptionalPrimitiveType will generate a direct assignment if both types
+// have the same underlying primitive type and have the same optionality
+//
+// if <source> != nil {
+//    <destination> = *<source>
+// } else {
+//    <destination> = <zero>
+// }
+func assignPrimitiveTypeFromOptionalPrimitiveType(
+	sourceEndpoint *StorageConversionEndpoint,
+	destinationEndpoint *StorageConversionEndpoint) StorageTypeConversion {
+
+	srcOpt := AsOptionalType(sourceEndpoint.Type())
+	dstOpt := AsOptionalType(destinationEndpoint.Type())
+
+	if srcOpt == nil || dstOpt != nil {
+		// Require source to be optional and destination to be non-optional
+		return nil
+	}
+
+	srcPrim := AsPrimitiveType(srcOpt)
+	dstPrim := AsPrimitiveType(destinationEndpoint.Type())
+
+	if srcPrim == nil || dstPrim == nil || !srcPrim.Equals(dstPrim) {
+		// Either or both sides are not primitive types, or not the same primitive type
+		return nil
+	}
+
+	return func(reader dst.Expr, writer dst.Expr, ctx *CodeGenerationContext) []dst.Stmt {
+		// Need to check for null and only assign if we have a value
+		cond := astbuilder.NotEqual(reader, dst.NewIdent("nil"))
+
+		assignValue := astbuilder.SimpleAssignment(writer, token.ASSIGN, astbuilder.Dereference(reader))
+
+		assignZero := astbuilder.SimpleAssignment(
+			writer,
+			token.ASSIGN,
+			&dst.BasicLit{
+				Value: zeroValue(srcPrim),
+			})
+
+		stmt := &dst.IfStmt{
+			Cond: cond,
+			Body: &dst.BlockStmt{
+				List: []dst.Stmt{
+					assignValue,
+				},
+			},
+			Else: &dst.BlockStmt{
+				List: []dst.Stmt{
+					assignZero,
+				},
+			},
+		}
+
+		return []dst.Stmt{stmt}
+	}
+}
+
+// assignArrayFromArray will generate a code fragment to populate an array, assuming the
+// underlying types of the two arrays are compatible
+//
+// <arr> := make([]<type>, len(<reader>))
+// for <index>, <value> := range <reader> {
+//     <arr>[<index>] := <value> // Or other conversion as required
+// }
+// <writer> = <arr>
+//
+func assignArrayFromArray(
+	sourceEndpoint *StorageConversionEndpoint,
+	destinationEndpoint *StorageConversionEndpoint) StorageTypeConversion {
+	st := AsArrayType(sourceEndpoint.Type())
+	dt := AsArrayType(destinationEndpoint.Type())
+
+	if st == nil || dt == nil {
+		// One or other type is not an array
+		return nil
+	}
+
+	srcEp := sourceEndpoint.WithType(st.element)
+	dstEp := destinationEndpoint.WithType(dt.element)
+	conversion, _ := createTypeConversion(srcEp, dstEp)
```

Comment:
minor: a comment explaining that this error is ignored because the intent of this function is to return nil if no conversion matches, and `createTypeConversion` returns `err` in that case might be worth it?

Reply:
Added at the start of the block